#Script to extract information on likely transposon insertion sites
#Edited 17/11/17

library(dplyr)
library(ggplot2)

setwd("~/Documents/work/jic/sulf/sulf_transposon_mapping/data")


#
# Read data ----
#
#Read in inputs generated by "reading_paired_unpaired.R" script
mut <- read.table("mutant_counts.tsv")
wt <- read.table("WT_counts.tsv")

#Add column names
colnames(wt) <- c("row", "chr","pos","depth","pairedF","pairedR","unpairF","unpairR")
colnames(mut) <- c("row", "chr","pos","depth","pairedF","pairedR","unpairF","unpairR")

# Add an ID to each
wt$genotype <- "WT"
mut$genotype <- "mutant"

# Bind the two tables together
sulf <- bind_rows(wt, mut)
rm(wt, mut)



#
# Apply filters to identify putative transposon insertions ----
#
# Keep sites with putative insertions, which include sites with either:
## many unpairedF + pairedR
## or
## many unpairedR + pairedF
sulf_filter <- filter(sulf, 
                      (unpairF > 2 & pairedF < 2 & unpairR < 2 & pairedR > 2) | 
                        (unpairF<2 & pairedF>2 & unpairR>2 & pairedR<2))

# Count how many there are in each genotype
## Roughly the same number - many insertions must be common between the two
count(sulf_filter, genotype)

# Retain only sites exclusive to the mutant genotype
## This reduces the number of sites a lot
sulf_specific <- sulf_filter %>% 
  group_by(chr, pos) %>% 
  filter(genotype == "mutant" & length(genotype) == 1) %>% 
  ungroup()

#Exclude sites within 100bp of the start/end of scaffolds 
## these will inevitably have an excess of unpaired counts
sulf_specific <- sulf_specific %>% 
  group_by(chr) %>% 
  filter(pos >= 100 & pos <= max(pos)-100)

# Classify sites as to which strand is missing more pairs
sulf_specific <- sulf_specific %>% 
  mutate(fwd_strand = unpairF > unpairR)

# Find inversion points - cases where strand switches from "Fwd" to "Rev"
## We use a trick to subtract the TRUE/FALSE fwd_strand from itself shifted by 1 position
## cases where this subtraction == 1 are the inversion points
## See for example:
## c(1, 1, 0, 0, 1, 1, 0, 0) - lead(c(1, 1, 0, 0, 1, 1, 0, 0))
## We also find the distance between the two sites where this "inversion" occurs
sulf_insertions <- sulf_specific %>% 
  group_by(chr) %>% 
  mutate(inversion = fwd_strand - lead(fwd_strand),
         inversion_length = lead(pos) - pos) %>% 
  filter(inversion == 1)

# Save this list of candidate insertions
write.csv(sulf_insertions, 'sulfSpecific.csv', row.names = F, quote = F)


#
# Visualisation ----
#
# This would show the distribution of inversion lengths (only few insertions in example)
ggplot(sulf_insertions, aes(inversion_length)) +
  geom_histogram()

# This shows their location along the genome
# Once all chromosomes are available, this might resemble a "map" of candidate insertions
# The "nudge_x" parameter might need adjusting
ggplot(sulf_insertions, aes(chr, pos/1e6)) +
  geom_point(size = 10, shape = "_") +
  geom_vline(aes(xintercept = as.numeric(chr))) +
  geom_label(aes(label = paste0(round(inversion_length/1e3, 2), " Kb")), nudge_x = 0.05) +
  labs(x = "Chromosome", y = "Position (Mb)") +
  theme_classic()

